### PR TITLE
Implemented client-side per-session and per-statement configurable timeouts

### PR DESCRIPTION
**Pull request by wprzytula**
Two levels of client-side timeouts were added: per Session and per Statement (PreparedStatement and Query). If no timeout is specified for a statement, a global per-session one is used. Once a timeout elapses, an `RequestTimeout` error is returned.

Fixes: #304

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by wprzytula** [comment]
@psarna @cvybhu Maybe you could help me with understanding why `test_client_timeout` fails in nondeterministic way.

**Comment by cvybhu** [comment]
> @psarna @cvybhu Maybe you could help me with understanding why `test_client_timeout` fails in nondeterministic way.

`tokio::Timeout` checks if the future has completed before checking if the timeout has elapsed.
With a local node and fast network it's possible that the query completes in a single `Poll` maybe?

https://github.com/tokio-rs/tokio/blob/de81985762a242c77361a6ab9de198372ca85987/tokio/src/time/timeout.rs#L175

**Comment by cvybhu** [comment]
Or maybe not, I added `tokio::task::yield_now().await;` at the beginning of runner block and it still fails.
So there are at least two Polls of the timeout future and it doesn't fail 0_o

**Comment by cvybhu** [comment]
Yeah something like this fails as well:
```rust
async fn yields_once() {
    tokio::task::yield_now().await;
}

async fn instant_timeout_test() {
    let res = tokio::time::timeout(std::time::Duration::from_secs(0), yields_once()).await;
    assert!(res.is_err());
}
```

Although sleeping for 1ms is enough to trigger timeout.

This might be an issue with tokio, but I'm not sure if it's a bug or a feature.
Tokio time keeps mentioning that its precision can be > 1ms, for example for `sleep`:

> The implementation is platform specific, and some platforms (specifically Windows) will provide timers with a larger resolution than 1 ms.

https://docs.rs/tokio/latest/tokio/time/fn.sleep.html

**Comment by cvybhu** [comment]
> Tokio time keeps mentioning that its precision can be > 1ms, for example for sleep:

I think we can just assume that tokio doesn't support timeouts of < 1ms and wait with the tests until proxy is merged.

**Comment by wprzytula** [comment]
> > Tokio time keeps mentioning that its precision can be > 1ms, for example for sleep:
> 
> I think we can just assume that tokio doesn't support timeouts of < 1ms and wait with the tests until proxy is merged.

That's right, with proxy we will easily delay the response for a couple of millis and set the timeout for 1 ms - this should work.

File: scylla-cql/src/errors.rs [hunk 1]
```diff
@@ -40,6 +40,10 @@ pub enum QueryError {
 
     #[error("Unable to allocate stream id")]
     UnableToAllocStreamId,
+
+    /// Client timeout occurred before any response arrived
+    #[error("Client timeout: {0}")]
+    ClientTimeout(String),
```
Comment (psarna) [inline scylla-cql/src/errors.rs:46]: In at least one other driver (Python) this thing is called request_timeout, so let's call it that. It is a client-side timeout indeed, but request_timeout is more descriptive

**Comment by psarna** [comment]
@wprzytula for now, let's just verify manually. If you create a 0ms or 1ms timeout, it should properly time out when sent to Scylla booted on a remote machine (e.g. one of the office servers).

**Comment by wprzytula** [comment]
v2: 
- client_timeout -> request_timeout
- timeout test made working with 1ms timeout and remote Scylla; test marked as ignored (for normal testing with local Scylla).

File: scylla-cql/src/errors.rs [hunk 2]
```diff
@@ -40,6 +40,10 @@ pub enum QueryError {
 
     #[error("Unable to allocate stream id")]
     UnableToAllocStreamId,
+
+    /// Client timeout occurred before any response arrived
+    #[error("Request timeout: {0}")]
+    RequestTimeout(String),
```
Comment (piodul) [inline scylla-cql/src/errors.rs:46]: Hmm, it looks like we will have two "timeout" errors now... Maybe we could provide a helper method `is_timeout()` which tells whether this error is a timeout (either a client- or server-side timeout)?

File: scylla/src/statement/mod.rs [hunk 3]
```diff
@@ -21,6 +21,7 @@ pub struct StatementConfig {
 
     pub tracing: bool,
     pub timestamp: Option<i64>,
+    pub request_timeout: Option<std::time::Duration>,
```
Comment (piodul) [inline scylla/src/statement/mod.rs:24]: Nit: how about a `use` for `std::time::Duration`?

File: scylla/src/transport/session_test.rs [hunk 4]
```diff
@@ -1181,6 +1181,78 @@ async fn test_timestamp() {
...
+    {
+        let session = SessionBuilder::new()
+            .known_node(uri.as_str())
+            .build()
+            .await
+            .unwrap();
+        session
+            .await_timed_schema_agreement(Duration::from_secs(1))
+            .await
+            .unwrap();
```
Comment (piodul) [inline scylla/src/transport/session_test.rs:1199]: Nit: waiting for schema agreement explicitly is no longer required, by default the driver now waits for schema agreement after every schema-altering query.

File: scylla/src/statement/prepared_statement.rs [hunk 5]
```diff
@@ -247,6 +248,19 @@ impl PreparedStatement {
         self.config.timestamp
     }
 
+    /// Sets the client-side timeout for this statement.
+    /// If not None, the driver will stop waiting for the request
+    /// to finish after `timeout` passed.
+    /// Otherwise, default session client timeout will be applied.
+    pub fn set_request_timeout(&mut self, timeout: Option<Duration>) {
```
Comment (piodul) [inline scylla/src/statement/prepared_statement.rs:255]: Nit: `timeout: None` can be easily misinterpreted as "no timeout". I don't have an idea for a better alternative, though :(

Comment (wprzytula) [reply scylla/src/statement/prepared_statement.rs:255]: In discussion with @psarna, this behaviour was chosen as appropriate.

Comment (psarna) [reply scylla/src/statement/prepared_statement.rs:255]: yup - `None` here means "not specified timeout", hopefully nobody would get confused by that. An an unlimited timeout, which isn't exactly recommended, can be simulated by setting a multi-hour timeout if need be

**Review by piodul — APPROVED** [review]

**Comment by wprzytula** [comment]
@psarna This is probably complete.

**Review by psarna — APPROVED** [review]

**Pull request merged**